backends/ebpf/ebpfstf.py

```diff
@@ -0,0 +1,123 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Contains functions which allow the parsing of stf files
+    and ultimately generate a C header file which defines the control plane
+    commands for that particular P4 program. """
+
+
+import os
+import sys
+from ebpfutils import *
+sys.path.insert(0, os.path.dirname(__file__) + '/../../tools/stf')
+from stf_parser import STFParser
+
+
+def _generate_control_actions(actions):
+    """ Generates the actual control plane commands.
+    This function inserts C code for all the "add" commands that have
+    been parsed. """
+    generated = ""
+    for index, cmd in enumerate(actions):
+        key_name = "key_%s%d" % (cmd["table"], index)
+        value_name = "value_%s%d" % (cmd["table"], index)
+        if cmd["type"] == "setdefault":
+            tbl_name = cmd["table"] + "_defaultAction"
+        else:
+            tbl_name = cmd["table"]
+        generated = "struct %s_key %s = {};\n\t" % (cmd["table"], key_name)
+        for key_num, key_field in enumerate(cmd["match"]):
+            field = key_field[0].split('.')[1]
+            generated += ("%s.%s = %s;\n\t"
+                          % (key_name, field, key_field[1]))
+        generated += ("int tableFileDescriptor = "
+                      "BPF_OBJ_GET(MAP_PATH \"/%s\");\n\t" %
+                      tbl_name)
+        generated += ("if (tableFileDescriptor < 0) {"
+                      "fprintf(stderr, \"map %s not loaded\");"
+                      " exit(1); }\n\t" % tbl_name)
+        generated += ("struct %s_value %s = {\n\t\t" % (
+            cmd["table"], value_name))
+        generated += ".action = %s,\n\t\t" % (cmd["action"][0])
+        if cmd["action"][1]:
+            generated += ".u = {"
+            for val_num, val_field in enumerate(cmd["action"][1]):
+                generated += "%s," % val_field[1]
+            generated += "},\n\t"
+        generated += "};\n\t"
+        generated += ("ok = BPF_USER_MAP_UPDATE_ELEM"
+                      "(tableFileDescriptor, &%s, &%s, BPF_ANY);\n\t"
+                      % (key_name, value_name))
+        generated += ("if (ok != 0) { perror(\"Could not write in %s\");"
+                      "exit(1); }\n" % tbl_name)
+    return generated
+
+
+def create_table_file(actions, tmpdir):
+    """ Create the control plane file.
+    The control commands are provided by the stf parser.
+    This generated file is required by ebpf_runtime.c to initialize
+    the control plane. """
+    control_file = open(tmpdir + "/control.h", "w+")
```

Comment:
Can you have the file name as a parameter that is passed from the caller?
I think you should also close the file; perhaps use a with statement